File: backend/gateway/src/lib/dto.ts
// Serializadores fila → DTO REST (specs/openapi.yaml). Los mismos DTO se usan
// en el WebSocket (los payloads del WS son las formas REST, ADR-IMPL-08).
// Convenciones pg: uuid/bigint/numeric → string, int → number, timestamptz → Date.
/* eslint-disable @typescript-eslint/no-explicit-any */
import { clean, iso } from './envelope.js';

type Row = Record<string, any>;

function num(v: unknown): number | undefined {
  if (v === null || v === undefined) return undefined;
  return Number(v);
}

function str(v: unknown): string | undefined {
  if (v === null || v === undefined) return undefined;
  return String(v);
}

function geo(v: unknown): unknown {
  if (v === null || v === undefined) return undefined;
  if (typeof v === 'string') return JSON.parse(v);
  return v;
}

export function accountDto(r: Row): Row {
  return clean({
    id: r.id,
    kind: r.kind,
    name: r.name,
    status: r.status,
    bot_archetype: r.kind === 'bot' ? (r.bot_archetype ?? undefined) : undefined,
    created_at: iso(r.created_at),
  });
}

export function ledgerAccountDto(r: Row): Row {
  return clean({
    id: r.id,
    kind: r.kind,
    owner_account_id: r.owner_account_id ?? undefined,
    product_id: r.product_id ?? undefined,
    warehouse_building_id: r.warehouse_building_id ?? undefined,
    reference_id: r.reference_id ?? undefined,
    balance: String(r.balance),
    updated_at: iso(r.updated_at),
    created_at: iso(r.created_at),
  });
}

export function ledgerEntryDto(r: Row): Row {
  return clean({
    id: r.id,
    transaction_id: r.transaction_id,
    account_id: r.account_id,
    amount: String(r.amount),
    transaction_kind: r.transaction_kind,
    reference_id: r.reference_id ?? undefined,
    description: r.description ?? undefined,
    sim_time_at: num(r.sim_time_at),
    created_at: iso(r.created_at),
  });
}

export function publicationDto(r: Row): Row {
  return clean({
    id: r.id,
    kind: r.kind,
    publisher_account_id: r.publisher_account_id,
    channel: r.channel,
    counterparty_account_id: r.counterparty_account_id ?? undefined,
    product_id: r.product_id ?? undefined,
    quantity_total: String(r.quantity_total),
    quantity_remaining: String(r.quantity_remaining),
    unit_price: String(r.unit_price),
    min_lot: String(r.min_lot),
    origin_node_id: r.origin_node_id ?? undefined,
    destination_node_id: r.destination_node_id ?? undefined,
    delivery_sim_seconds: num(r.delivery_sim_seconds),
    status: r.status,
    window_closes_at: iso(r.window_closes_at),
    cancel_cooldown_until: iso(r.cancel_cooldown_until),
    published_at_sim: num(r.published_at_sim),
    created_at: iso(r.created_at),
  });
}

export function acceptanceDto(r: Row): Row {
  return clean({
    id: r.id,
    publication_id: r.publication_id,
    acceptor_account_id: r.acceptor_account_id,
    quantity: String(r.quantity),
    quantity_served: String(r.quantity_served),
    status: r.status,
    draw_order: r.draw_order ?? undefined,
    contract_id: r.contract_id ?? undefined,
    freight_contract_id: r.freight_contract_id ?? undefined,
    accepted_at: iso(r.accepted_at),
    resolved_at: iso(r.resolved_at),
  });
}

export function contractDto(r: Row): Row {
  return clean({
    id: r.id,
    publication_id: r.publication_id ?? undefined,
    channel: r.channel,
    buyer_account_id: r.buyer_account_id,
    seller_account_id: r.seller_account_id,
    product_id: r.product_id,
    quantity_agreed: String(r.quantity_agreed),
    quantity_delivered: String(r.quantity_delivered),
    unit_price: String(r.unit_price),
    origin_node_id: r.origin_node_id,
    destination_node_id: r.destination_node_id,
    deadline_sim: num(r.deadline_sim),
    status: r.status,
    fill_bp: r.fill_bp ?? undefined,
    stock_reserve_account_id: r.stock_reserve_account_id ?? undefined,
    seller_guarantee_account_id: r.seller_guarantee_account_id ?? undefined,
    escrow_account_id: r.escrow_account_id ?? undefined,
    confirmed_at_sim: num(r.confirmed_at_sim),
    settled_at_sim: num(r.settled_at_sim),
    created_at: iso(r.created_at),
  });
}

export function contractDeliveryDto(r: Row): Row {
  return clean({
    id: r.id,
    contract_id: r.contract_id,
    shipment_id: r.shipment_id,
    quantity: String(r.quantity),
    delivered_at_sim: num(r.delivered_at_sim),
    on_time: r.on_time,
  });
}

export function regionDto(r: Row): Row {
  return clean({
    id: r.id,
    name: r.name,
    grid_x: r.grid_x,
    grid_y: r.grid_y,
    bounds: geo(r.bounds_geojson),
    biome: r.biome,
    tax_rate_bp: r.tax_rate_bp,
    customs_rate_bp: r.customs_rate_bp,
    canon_base: String(r.canon_base),
    opened_at_sim: num(r.opened_at_sim),
  });
}

export function productDto(r: Row): Row {
  return clean({
    id: r.id,
    code: r.code,
    name: r.name,
    class: r.class,
    unit_volume: r.unit_volume,
    base_price: String(r.base_price),
    price_floor: String(r.price_floor),
    price_ceiling: String(r.price_ceiling),
    is_fuel: r.is_fuel,
  });
}

export function buildingTypeDto(r: Row): Row {
  return clean({
    id: r.id,
    code: r.code,
    name: r.name,
    footprint_cells: r.footprint_cells,
    max_level: r.max_level,
    base_storage: String(r.base_storage),
    placement_rules: r.placement_rules,
    level_curve: r.level_curve,
    build_cost: String(r.build_cost),
    maintenance_cost: String(r.maintenance_cost),
  });
}

export function recipeDto(r: Row): Row {
  const ingredients = Array.isArray(r.ingredients) ? r.ingredients : [];
  return clean({
    id: r.id,
    building_type_id: r.building_type_id,
    code: r.code,
    name: r.name,
    batch_sim_seconds: num(r.batch_sim_seconds),
    fuel_product_id: r.fuel_product_id ?? undefined,
    fuel_per_batch: String(r.fuel_per_batch),
    workers_required: r.workers_required,
    min_city_level: r.min_city_level,
    changeover_seconds: num(r.changeover_seconds),
    ingredients: ingredients.map((i: Row) =>
      clean({ product_id: i.product_id, role: i.role, quantity: String(i.quantity) }),
    ),
  });
}

export function depositDto(r: Row): Row {
  return clean({
    id: r.id,
    region_id: r.region_id,
    product_id: r.product_id,
    location: geo(r.location_geojson),
    initial_amount: String(r.initial_amount),
    remaining_amount: String(r.remaining_amount),
    renewable: r.renewable,
    regen_per_sim_day: String(r.regen_per_sim_day),
  });
}

export function cityDto(r: Row): Row {
  return clean({
    id: r.id,
    region_id: r.region_id,
    account_id: r.account_id,
    name: r.name,
    location: geo(r.location_geojson),
    level: r.level,
    population: num(r.population),
    supply_index: Number(r.supply_index),
    influence_radius_m: r.influence_radius_m,
    base_salary: String(r.base_salary),
  });
}

export function cityDemandDto(r: Row): Row {
  return clean({
    city_id: r.city_id,
    product_id: r.product_id,
    d0_per_sim_day: String(r.d0_per_sim_day),
    saturation_factor: Number(r.saturation_factor),
    current_price: String(r.current_price),
    unlocked_at_level: r.unlocked_at_level,
    updated_at_sim: num(r.updated_at_sim),
  });
}

export function concessionDto(r: Row): Row {
  return clean({
    id: r.id,
    region_id: r.region_id,
    holder_account_id: r.holder_account_id,
    parcel: geo(r.parcel_geojson),
    canon_amount: String(r.canon_amount),
    period_sim_days: r.period_sim_days,
    expires_at_sim: num(r.expires_at_sim),
    status: r.status,
    granted_at_sim: num(r.granted_at_sim),
  });
}

export function concessionTransferDto(r: Row): Row {
  return clean({
    id: r.id,
    concession_id: r.concession_id,
    from_account_id: r.from_account_id,
    to_account_id: r.to_account_id,
    price: String(r.price),
    system_fee: String(r.system_fee),
    occurred_at_sim: num(r.occurred_at_sim),
  });
}

export function buildingDto(r: Row): Row {
  return clean({
    id: r.id,
    owner_account_id: r.owner_account_id,
    region_id: r.region_id,
    concession_id: r.concession_id,
    building_type_id: r.building_type_id,
    footprint: geo(r.footprint_geojson),
    level: r.level,
    status: r.status,
    active_recipe_id: r.active_recipe_id ?? undefined,
    condition_pct: r.condition_pct,
    fuel_stock: String(r.fuel_stock),
    updated_at_sim: num(r.updated_at_sim),
    created_at: iso(r.created_at),
  });
}

export function inventoryItemDto(r: Row): Row {
  return clean({
    building_id: r.building_id,
    product_id: r.product_id,
    quantity: String(r.quantity),
    updated_at_sim: num(r.updated_at_sim),
  });
}

/** El progreso del lote es analítico: derivado en el momento de la consulta. */
export function productionBatchDto(r: Row, simSeconds: number, batchSimSeconds?: number): Row {
  let progress: number | undefined;
  let eta: number | undefined;
  if (r.status === 'running' && r.started_at_sim !== null && batchSimSeconds && batchSimSeconds > 0) {
    const elapsed = simSeconds - Number(r.started_at_sim);
    progress = Math.min(100, Math.max(0, (100 * elapsed) / batchSimSeconds));
    eta = Number(r.started_at_sim) + batchSimSeconds;
  }
  return clean({
    id: r.id,
    building_id: r.building_id,
    recipe_id: r.recipe_id,
    batches_queued: r.batches_queued,
    batches_done: r.batches_done,
    status: r.status,
    queue_position: r.queue_position,
    started_at_sim: num(r.started_at_sim),
    progress_pct: progress,
    eta_sim: eta,
  });
}

export function vehicleTypeDto(r: Row): Row {
  return clean({
    id: r.id,
    code: r.code,
    name: r.name,
    mode: r.mode,
    cargo_capacity: String(r.cargo_capacity),
    speed_kmh: r.speed_kmh,
    fuel_product_id: r.fuel_product_id,
    fuel_per_100km: String(r.fuel_per_100km),
    autonomy_km: r.autonomy_km,
    purchase_price: String(r.purchase_price),
    operating_cost_per_day: String(r.operating_cost_per_day),
  });
}

/**
 * SELECT de vehículos con posición derivada. Requiere $N = sim_seconds como
 * ÚLTIMO parámetro del query (índice `simParamIdx`).
 * Posición analítica: progreso = (sim_now − segment_entered_sim) / duration.
 */
export function vehicleSelectSql(simParam: string): string {
  return `
    SELECT v.id, v.vehicle_type_id, v.owner_account_id, v.status, v.wear_pct, v.fuel,
           v.route_id, v.route_leg_index, v.repair_until_sim, v.updated_at_sim,
           v.at_node_id, v.on_segment_id, v.segment_entered_sim,
           (v.advance_fn->>'duration_sim_seconds')::float8 AS seg_duration,
           CASE WHEN v.at_node_id IS NOT NULL THEN ST_AsGeoJSON(n.location) END AS node_loc,
           CASE WHEN v.on_segment_id IS NOT NULL THEN ST_AsGeoJSON(
             ST_LineInterpolatePoint(s.portion, LEAST(1.0, GREATEST(0.0,
               (${simParam}::float8 - v.segment_entered_sim::float8)
                 / NULLIF((v.advance_fn->>'duration_sim_seconds')::float8, 0)
             )))) END AS seg_loc
      FROM world.vehicles v
      LEFT JOIN world.network_nodes n ON n.id = v.at_node_id
      LEFT JOIN world.link_segments s ON s.id = v.on_segment_id`;
}

export function vehicleDto(r: Row, simSeconds: number): Row {
  const position: Row = {};
  if (r.at_node_id) {
    position.at_node_id = r.at_node_id;
    const loc = geo(r.node_loc);
    if (loc) position.location = loc;
  } else if (r.on_segment_id) {
    position.on_segment_id = r.on_segment_id;
    const dur = Number(r.seg_duration);
    if (Number.isFinite(dur) && dur > 0 && r.segment_entered_sim !== null) {
      const pct = (100 * (simSeconds - Number(r.segment_entered_sim))) / dur;
      position.segment_progress_pct = Math.min(100, Math.max(0, pct));
    }
    const loc = geo(r.seg_loc);
    if (loc) position.location = loc;
  }
  return clean({
    id: r.id,
    vehicle_type_id: r.vehicle_type_id,
    owner_account_id: r.owner_account_id,
    status: r.status,
    wear_pct: r.wear_pct,
    fuel: String(r.fuel),
    route_id: r.route_id ?? undefined,
    route_leg_index: r.route_leg_index ?? undefined,
    position,
    repair_until_sim: num(r.repair_until_sim),
    updated_at_sim: num(r.updated_at_sim),
  });
}

export function shipmentDto(r: Row): Row {
  return clean({
    id: r.id,
    owner_account_id: r.owner_account_id,
    product_id: r.product_id,
    quantity: String(r.quantity),
    contract_id: r.contract_id ?? undefined,
    freight_contract_id: r.freight_contract_id ?? undefined,
    vehicle_id: r.vehicle_id ?? undefined,
    at_node_id: r.at_node_id ?? undefined,
    status: r.status,
    updated_at_sim: num(r.updated_at_sim),
  });
}

export function terminalDto(r: Row): Row {
  return clean({
    id: r.id,
    node_id: r.node_id,
    owner_account_id: r.owner_account_id,
    transshipment_per_hour: r.transshipment_per_hour,
    queue_length: r.queue_length,
    updated_at_sim: num(r.updated_at_sim),
  });
}

export function terminalSlotDto(r: Row): Row {
  return clean({
    id: r.id,
    terminal_id: r.terminal_id,
    priority_tier: r.priority_tier,
    price: String(r.price),
    holder_account_id: r.holder_account_id ?? undefined,
    valid_until_sim: num(r.valid_until_sim),
  });
}

export function networkNodeDto(r: Row): Row {
  return clean({
    id: r.id,
    kind: r.kind,
    region_id: r.region_id,
    building_id: r.building_id ?? undefined,
    city_id: r.city_id ?? undefined,
    location: geo(r.location_geojson),
  });
}

export function networkLinkDto(r: Row): Row {
  const segments = Array.isArray(r.segments) ? r.segments : [];
  return clean({
    id: r.id,
    mode: r.mode,
    from_node_id: r.from_node_id,
    to_node_id: r.to_node_id,
    path: geo(r.path_geojson),
    length_m: r.length_m,
    capacity_per_hour: r.capacity_per_hour,
    base_speed_kmh: r.base_speed_kmh,
    segments: segments.map((s: Row) =>
      clean({
        id: s.id,
        region_id: s.region_id,
        seq: s.seq,
        length_m: s.length_m,
        congestion_ema: Number(s.congestion_ema),
        updated_at_sim: num(s.updated_at_sim),
      }),
    ),
  });
}

export function routeDto(r: Row): Row {
  const legs = Array.isArray(r.legs) ? r.legs : [];
  return clean({
    id: r.id,
    owner_account_id: r.owner_account_id,
    name: r.name,
    kind: r.kind,
    active: r.active,
    legs: legs.map((l: Row) => ({ leg_index: l.leg_index, link_id: l.link_id })),
    created_at: iso(r.created_at),
  });
}

export function ohlcDto(r: Row, productId: string, bucketSimSecs: number): Row {
  return clean({
    product_id: productId,
    region_id: r.region_id,
    bucket_start_sim: num(r.bucket_start_sim),
    bucket_sim_secs: bucketSimSecs,
    open_price: String(r.open_price),
    high_price: String(r.high_price),
    low_price: String(r.low_price),
    close_price: String(r.close_price),
    volume: String(r.volume),
    contract_count: Number(r.contract_count),
  });
}

File: backend/engine/internal/outbox/dto.go
package outbox

// Constructores de DTOs REST (specs/openapi.yaml) para los payloads de eventos.
// Convenciones transversales: dinero/stock como strings decimales, sim-time
// como enteros, timestamps RFC3339, geometrías GeoJSON.

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"

	"imperio/engine/internal/core"
)

func fmtTime(t time.Time) string { return t.UTC().Format(time.RFC3339) }

func put(m map[string]any, key string, v any) {
	// Añade solo valores no nulos (los DTO omiten los opcionales ausentes).
	switch x := v.(type) {
	case *uuid.UUID:
		if x != nil {
			m[key] = x.String()
		}
	case *int64:
		if x != nil {
			m[key] = *x
		}
	case *int32:
		if x != nil {
			m[key] = *x
		}
	case *string:
		if x != nil {
			m[key] = *x
		}
	case *time.Time:
		if x != nil {
			m[key] = fmtTime(*x)
		}
	default:
		if v != nil {
			m[key] = v
		}
	}
}

// PublicationEntity — DTO Publication.
func PublicationEntity(ctx context.Context, q core.Querier, id uuid.UUID) (map[string]any, error) {
	var (
		kind, channel, status                     string
		publisher                                 uuid.UUID
		counterparty, productID, originID, destID *uuid.UUID
		qtyTotal, qtyRemaining, unitPrice, minLot int64
		deliverySimSeconds, publishedAtSim        int64
		windowClosesAt, cancelCooldownUntil       *time.Time
		createdAt                                 time.Time
	)
	err := q.QueryRow(ctx, `
		SELECT kind, publisher_account_id, channel, counterparty_account_id, product_id,
		       quantity_total, quantity_remaining, unit_price, min_lot,
		       origin_node_id, destination_node_id, delivery_sim_seconds, status,
		       window_closes_at, cancel_cooldown_until, published_at_sim, created_at
		  FROM ledger.publications WHERE id = $1`, id).Scan(
		&kind, &publisher, &channel, &counterparty, &productID,
		&qtyTotal, &qtyRemaining, &unitPrice, &minLot,
		&originID, &destID, &deliverySimSeconds, &status,
		&windowClosesAt, &cancelCooldownUntil, &publishedAtSim, &createdAt)
	if err != nil {
		return nil, fmt.Errorf("dto publication %s: %w", id, err)
	}
	m := map[string]any{
		"id":                   id.String(),
		"kind":                 kind,
		"publisher_account_id": publisher.String(),
		"channel":              channel,
		"quantity_total":       core.Money(qtyTotal),
		"quantity_remaining":   core.Money(qtyRemaining),
		"unit_price":           core.Money(unitPrice),
		"min_lot":              core.Money(minLot),
		"delivery_sim_seconds": deliverySimSeconds,
		"status":               status,
		"published_at_sim":     publishedAtSim,
		"created_at":           fmtTime(createdAt),
	}
	put(m, "counterparty_account_id", counterparty)
	put(m, "product_id", productID)
	put(m, "origin_node_id", originID)
	put(m, "destination_node_id", destID)
	put(m, "window_closes_at", windowClosesAt)
	put(m, "cancel_cooldown_until", cancelCooldownUntil)
	return m, nil
}

// ContractEntity — DTO Contract.
func ContractEntity(ctx context.Context, q core.Querier, id uuid.UUID) (map[string]any, error) {
	var (
		publicationID                      *uuid.UUID
		channel, status                    string
		buyer, seller, product             uuid.UUID
		origin, dest                       uuid.UUID
		qtyAgreed, qtyDelivered, unitPrice int64
		deadlineSim, confirmedAtSim        int64
		fillBp                             *int32
		settledAtSim                       *int64
		stockAcc, guarAcc, escrowAcc       uuid.UUID
		createdAt                          time.Time
	)
	err := q.QueryRow(ctx, `
		SELECT publication_id, channel, buyer_account_id, seller_account_id, product_id,
		       quantity_agreed, quantity_delivered, unit_price,
		       origin_node_id, destination_node_id, deadline_sim, status, fill_bp,
		       stock_reserve_account_id, seller_guarantee_account_id, escrow_account_id,
		       confirmed_at_sim, settled_at_sim, created_at
		  FROM ledger.contracts WHERE id = $1`, id).Scan(
		&publicationID, &channel, &buyer, &seller, &product,
		&qtyAgreed, &qtyDelivered, &unitPrice,
		&origin, &dest, &deadlineSim, &status, &fillBp,
		&stockAcc, &guarAcc, &escrowAcc,
		&confirmedAtSim, &settledAtSim, &createdAt)
	if err != nil {
		return nil, fmt.Errorf("dto contract %s: %w", id, err)
	}
	m := map[string]any{
		"id":                          id.String(),
		"channel":                     channel,
		"buyer_account_id":            buyer.String(),
		"seller_account_id":           seller.String(),
		"product_id":                  product.String(),
		"quantity_agreed":             core.Money(qtyAgreed),
		"quantity_delivered":          core.Money(qtyDelivered),
		"unit_price":                  core.Money(unitPrice),
		"origin_node_id":              origin.String(),
		"destination_node_id":         dest.String(),
		"deadline_sim":                deadlineSim,
		"status":                      status,
		"stock_reserve_account_id":    stockAcc.String(),
		"seller_guarantee_account_id": guarAcc.String(),
		"escrow_account_id":           escrowAcc.String(),
		"confirmed_at_sim":            confirmedAtSim,
		"created_at":                  fmtTime(createdAt),
	}
	put(m, "publication_id", publicationID)
	put(m, "fill_bp", fillBp)
	put(m, "settled_at_sim", settledAtSim)
	return m, nil
}

// AcceptanceEntity — DTO Acceptance (contract_id se pasa aparte porque la
// tabla no lo persiste; lo conoce el sorteo que acaba de crear el contrato).
func AcceptanceEntity(ctx context.Context, q core.Querier, id uuid.UUID, contractID *uuid.UUID) (map[string]any, error) {
	var (
		publicationID, acceptor uuid.UUID
		quantity, served        int64
		status                  string
		drawOrder               *int32
		acceptedAt              time.Time
		resolvedAt              *time.Time
	)
	err := q.QueryRow(ctx, `
		SELECT publication_id, acceptor_account_id, quantity, quantity_served, status,
		       draw_order, accepted_at, resolved_at
		  FROM ledger.publication_acceptances WHERE id = $1`, id).Scan(
		&publicationID, &acceptor, &quantity, &served, &status, &drawOrder, &acceptedAt, &resolvedAt)
	if err != nil {
		return nil, fmt.Errorf("dto acceptance %s: %w", id, err)
	}
	m := map[string]any{
		"id":                  id.String(),
		"publication_id":      publicationID.String(),
		"acceptor_account_id": acceptor.String(),
		"quantity":            core.Money(quantity),
		"quantity_served":     core.Money(served),
		"status":              status,
		"accepted_at":         fmtTime(acceptedAt),
	}
	put(m, "draw_order", drawOrder)
	put(m, "resolved_at", resolvedAt)
	put(m, "contract_id", contractID)
	return m, nil
}

// BatchEntity — DTO ProductionBatch.
func BatchEntity(ctx context.Context, q core.Querier, id uuid.UUID) (map[string]any, error) {
	var (
		buildingID, recipeID uuid.UUID
		queued, done, pos    int32
		status               string
		startedAtSim         *int64
	)
	err := q.QueryRow(ctx, `
		SELECT building_id, recipe_id, batches_queued, batches_done, status, queue_position, started_at_sim
		  FROM world.production_batches WHERE id = $1`, id).Scan(
		&buildingID, &recipeID, &queued, &done, &status, &pos, &startedAtSim)
	if err != nil {
		return nil, fmt.Errorf("dto batch %s: %w", id, err)
	}
	m := map[string]any{
		"id":             id.String(),
		"building_id":    buildingID.String(),
		"recipe_id":      recipeID.String(),
		"batches_queued": queued,
		"batches_done":   done,
		"status":         status,
		"queue_position": pos,
	}
	put(m, "started_at_sim", startedAtSim)
	return m, nil
}

// BuildingEntity — DTO Building. Devuelve además el centroide (para la
// location de los eventos espaciales building.status_changed).
func BuildingEntity(ctx context.Context, q core.Querier, id uuid.UUID) (map[string]any, *Location, error) {
	var (
		owner, region, concession, btype uuid.UUID
		footprintJSON                    string
		level, conditionPct              int32
		status                           string
		activeRecipe                     *uuid.UUID
		fuelStock, updatedAtSim          int64
		createdAt                        time.Time
		lon, lat                         float64
	)
	err := q.QueryRow(ctx, `
		SELECT owner_account_id, region_id, concession_id, building_type_id,
		       ST_AsGeoJSON(footprint), level, status, active_recipe_id, condition_pct,
		       fuel_stock, updated_at_sim, created_at,
		       ST_X(ST_Centroid(footprint)), ST_Y(ST_Centroid(footprint))
		  FROM world.buildings WHERE id = $1`, id).Scan(
		&owner, &region, &concession, &btype,
		&footprintJSON, &level, &status, &activeRecipe, &conditionPct,
		&fuelStock, &updatedAtSim, &createdAt, &lon, &lat)
	if err != nil {
		return nil, nil, fmt.Errorf("dto building %s: %w", id, err)
	}
	m := map[string]any{
		"id":               id.String(),
		"owner_account_id": owner.String(),
		"region_id":        region.String(),
		"concession_id":    concession.String(),
		"building_type_id": btype.String(),
		"footprint":        json.RawMessage(footprintJSON),
		"level":            level,
		"status":           status,
		"condition_pct":    conditionPct,
		"fuel_stock":       core.Money(fuelStock),
		"updated_at_sim":   updatedAtSim,
		"created_at":       fmtTime(createdAt),
	}
	put(m, "active_recipe_id", activeRecipe)
	return m, &Location{Lon: lon, Lat: lat}, nil
}

// VehicleEntity — DTO Vehicle con posición derivada. simNow permite derivar
// el progreso del segmento (analítico, GDD 1.1). Devuelve la location.
func VehicleEntity(ctx context.Context, q core.Querier, id uuid.UUID, simNow int64) (map[string]any, *Location, error) {
	var (
		vtype, owner       uuid.UUID
		status             string
		wearPct            int32
		fuel, updatedAtSim int64
		routeID            *uuid.UUID
		routeLegIndex      *int32
		atNode, onSegment  *uuid.UUID
		segmentEnteredSim  *int64
		advanceFnRaw       []byte
		repairUntilSim     *int64
	)
	err := q.QueryRow(ctx, `
		SELECT vehicle_type_id, owner_account_id, status, wear_pct, fuel, route_id,
		       route_leg_index, at_node_id, on_segment_id, segment_entered_sim,
		       advance_fn, repair_until_sim, updated_at_sim
		  FROM world.vehicles WHERE id = $1`, id).Scan(
		&vtype, &owner, &status, &wearPct, &fuel, &routeID,
		&routeLegIndex, &atNode, &onSegment, &segmentEnteredSim,
		&advanceFnRaw, &repairUntilSim, &updatedAtSim)
	if err != nil {
		return nil, nil, fmt.Errorf("dto vehicle %s: %w", id, err)
	}

	position := map[string]any{}
	var loc *Location
	if atNode != nil {
		position["at_node_id"] = atNode.String()
		if l, err := NodeLocation(ctx, q, *atNode); err == nil {
			loc = l
			position["location"] = map[string]any{"type": "Point", "coordinates": []float64{l.Lon, l.Lat}}
		}
	} else if onSegment != nil {
		position["on_segment_id"] = onSegment.String()
		// Progreso analítico del segmento: (sim_now − t_entrada) / duración.
		progress := 0.0
		if segmentEnteredSim != nil && len(advanceFnRaw) > 0 {
			var fn core.AdvanceFn
			if json.Unmarshal(advanceFnRaw, &fn) == nil && fn.DurationSimSeconds > 0 {
				progress = float64(simNow-*segmentEnteredSim) / float64(fn.DurationSimSeconds)
			}
		}
		if progress < 0 {
			progress = 0
		}
		if progress > 1 {
			progress = 1
		}
		position["segment_progress_pct"] = progress * 100
		var lon, lat float64
		if err := q.QueryRow(ctx,
			`SELECT ST_X(ST_LineInterpolatePoint(portion, $2)), ST_Y(ST_LineInterpolatePoint(portion, $2))
			   FROM world.link_segments WHERE id = $1`, *onSegment, progress).Scan(&lon, &lat); err == nil {
			loc = &Location{Lon: lon, Lat: lat}
			position["location"] = map[string]any{"type": "Point", "coordinates": []float64{lon, lat}}
		}
	}

	m := map[string]any{
		"id":               id.String(),
		"vehicle_type_id":  vtype.String(),
		"owner_account_id": owner.String(),
		"status":           status,
		"wear_pct":         wearPct,
		"fuel":             core.Money(fuel),
		"position":         position,
		"updated_at_sim":   updatedAtSim,
	}
	put(m, "route_id", routeID)
	put(m, "route_leg_index", routeLegIndex)
	put(m, "repair_until_sim", repairUntilSim)
	return m, loc, nil
}

// CityEntity — DTO City (+ location para eventos espaciales city.*).
func CityEntity(ctx context.Context, q core.Querier, id uuid.UUID) (map[string]any, *Location, error) {
	var (
		region, account  uuid.UUID
		name             string
		level            int32
		population       int64
		supplyIndex      float64
		influenceRadiusM int32
		baseSalary       int64
		lon, lat         float64
	)
	err := q.QueryRow(ctx, `
		SELECT region_id, account_id, name, level, population, supply_index,
		       influence_radius_m, base_salary, ST_X(location), ST_Y(location)
		  FROM world.cities WHERE id = $1`, id).Scan(
		&region, &account, &name, &level, &population, &supplyIndex,
		&influenceRadiusM, &baseSalary, &lon, &lat)
	if err != nil {
		return nil, nil, fmt.Errorf("dto city %s: %w", id, err)
	}
	m := map[string]any{
		"id":                 id.String(),
		"region_id":          region.String(),
		"account_id":         account.String(),
		"name":               name,
		"location":           map[string]any{"type": "Point", "coordinates": []float64{lon, lat}},
		"level":              level,
		"population":         population,
		"supply_index":       supplyIndex,
		"influence_radius_m": influenceRadiusM,
		"base_salary":        core.Money(baseSalary),
	}
	return m, &Location{Lon: lon, Lat: lat}, nil
}

// CityDemandEntity — DTO CityDemand.
func CityDemandEntity(ctx context.Context, q core.Querier, cityID, productID uuid.UUID) (map[string]any, error) {
	var (
		d0           int64
		saturation   float64
		currentPrice int64
		unlockedAt   int32
		updatedAtSim int64
	)
	err := q.QueryRow(ctx, `
		SELECT d0_per_sim_day, saturation_factor, current_price, unlocked_at_level, updated_at_sim
		  FROM world.city_demand WHERE city_id = $1 AND product_id = $2`, cityID, productID).Scan(
		&d0, &saturation, &currentPrice, &unlockedAt, &updatedAtSim)
	if err != nil {
		return nil, fmt.Errorf("dto city_demand %s/%s: %w", cityID, productID, err)
	}
	return map[string]any{
		"city_id":           cityID.String(),
		"product_id":        productID.String(),
		"d0_per_sim_day":    core.Money(d0),
		"saturation_factor": saturation,
		"current_price":     core.Money(currentPrice),
		"unlocked_at_level": unlockedAt,
		"updated_at_sim":    updatedAtSim,
	}, nil
}
